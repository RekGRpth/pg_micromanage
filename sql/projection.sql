CREATE EXTENSION micromanage;

-- no range table -> error!
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
}
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');

CREATE TABLE a (a int);
INSERT INTO a VALUES (1);
INSERT INTO a VALUES (10);

SELECT * FROM run_select(:'buf');


-- SELECT d FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "d" } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- this is not expressable in sql, it should definitely fail!
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 2 column: "d" } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a, 10 FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  target: { const: { uint: 10 } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a + 10 FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: {
    op: {
      name: "+"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a, a > 5 FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  target: {
    op: {
      name: ">"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 5 }
      }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a FROM a WHERE a = 10;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: {
    op: {
      name: "="
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a FROM a WHERE a + 10;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: {
    op: {
      name: "+"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10; } }
    }
  }
}
rtable: { name: "a" } 
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a FROM a WHERE true;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: { const: { bool: true } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a FROM a WHERE false;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: { const: { bool: false } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a + true FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: {
    op: {
      name: "+"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { bool: true } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');


-- SELECT a = true FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: {
    op: {
      name: "="
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { bool: true } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');

CREATE TABLE b (a int);

-- adding unused range tables is okay
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
}
rtable: { name: "a" }
rtable: { name: "b" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');

-- SELECT a.a, b.a + 10 FROM a
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  target: {
    op: {
      name: "+"
      arg: { var: { table: 2 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
  qual: {
    op: {
      name: "="
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
rtable: { name: "b" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');

-- SELECT a.a FROM a WHERE b.a = 10
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: {
    op: {
      name: "="
      arg: { var: { table: 2 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
rtable: { name: "b" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');

DROP TABLE a;
DROP TABLE b;
