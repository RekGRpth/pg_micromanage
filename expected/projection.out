CREATE EXTENSION micromanage;
-- no range table -> error!
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
}
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
ERROR:  must include at least one range table
-- SELECT a FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
ERROR:  table a does not exist
CREATE TABLE a (a int);
INSERT INTO a VALUES (1);
INSERT INTO a VALUES (10);
SELECT * FROM run_select(:'buf');
 a  
----
  1
 10
(2 rows)

-- SELECT d FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "d" } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
ERROR:  could not find column d of table 1
-- this is not expressable in sql, it should definitely fail!
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 2 column: "d" } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
WARNING:  cant select from table 2, using table 1 instead
ERROR:  could not find column d of table 1
-- SELECT a, 10 FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  target: { const: { uint: 10 } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a  | a  
----+----
  1 | 10
 10 | 10
(2 rows)

-- SELECT a + 10 FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: {
    op: {
      name: "+"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a  
----
 11
 20
(2 rows)

-- SELECT a, a > 5 FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  target: {
    op: {
      name: ">"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 5 }
      }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a  | a 
----+---
  1 | f
 10 | t
(2 rows)

-- SELECT a FROM a WHERE a = 10;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: {
    op: {
      name: "="
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a  
----
 10
(1 row)

-- SELECT a FROM a WHERE a + 10;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: {
    op: {
      name: "+"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10; } }
    }
  }
}
rtable: { name: "a" } 
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
ERROR:  quals must return a boolean
-- SELECT a FROM a WHERE true;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: { const: { bool: true } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a  
----
  1
 10
(2 rows)

-- SELECT a FROM a WHERE false;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: { const: { bool: false } }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a 
---
(0 rows)

-- SELECT a + true FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: {
    op: {
      name: "+"
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { bool: true } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
ERROR:  could not find operator named "+"
DETAIL:  arg types: "int4" and "bool"
-- SELECT a = true FROM a;
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: {
    op: {
      name: "="
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { bool: true } }
    }
  }
}
rtable: { name: "a" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
ERROR:  could not find operator named "="
DETAIL:  arg types: "int4" and "bool"
CREATE TABLE b (a int);
-- adding unused range tables is okay
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
}
rtable: { name: "a" }
rtable: { name: "b" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
 a  
----
  1
 10
(2 rows)

-- SELECT a.a, b.a + 10 FROM a
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  target: {
    op: {
      name: "+"
      arg: { var: { table: 2 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
  qual: {
    op: {
      name: "="
      arg: { var: { table: 1 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
rtable: { name: "b" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
WARNING:  cant select from table 2, using table 1 instead
 a  | a  
----+----
 10 | 20
(1 row)

-- SELECT a.a FROM a WHERE b.a = 10
SELECT encode_protobuf($$
plan: {
  sscan: { table: 1 }
  target: { var: { table: 1 column: "a" } }
  qual: {
    op: {
      name: "="
      arg: { var: { table: 2 column: "a" } }
      arg: { const: { uint: 10 } }
    }
  }
}
rtable: { name: "a" }
rtable: { name: "b" }
$$) AS buf \gset
SELECT * FROM run_select(:'buf');
WARNING:  cant select from table 2, using table 1 instead
 a  
----
 10
(1 row)

DROP TABLE a;
DROP TABLE b;
